### PR TITLE
feat: commit and push changes onto a SHA to a target remote

### PR DESCRIPTION
**Pull request by TomKristie**
This PR implements the basic functionality of creating GitHub trees, a commit for that tree, and updating the latest reference HEAD to point to that tree. It also has basic mock tests.

Given:
- current reference HEAD
- an owner
- a repository
- a branch name
- changes
- commit message

On success, create a tree with those changes, a commit pointing to that tree, and update the reference branch HEAD to point to that new commit.
On error re-throw Octokit Error.

Towards #19

### CODE REVIEW

File: src/types/index.ts [hunk 1]
```diff
@@ -15,9 +15,27 @@
...
+  mode: FileMode;
+  content?: string;
+}
+
 // a flat object of the path of the file as the key, and the text contents as a value
-interface Files {
-  [index: string]: string;
+interface Changes {
+  [path: string]: FileData;
 }
```
Comment (chingor13) [inline src/types/index.ts:39]:
```suggestion
type Changes = Map<string, FileData>;
```

File: src/github-handler/commit-and-push-handler.ts [hunk 2]
```diff
@@ -0,0 +1,187 @@
...
+ * Generate and return a GitHub tree object structure
+ * containing the target change data
+ * See https://developer.github.com/v3/git/trees/#tree-object
+ * @param {Changes} changes the set of repository changes
+ * @returns {GitCreateTreeParamsTree[]} The new GitHub changes
+ */
+function genTreeObj(changes: Changes): GitCreateTreeParamsTree[] {
+  const tree: GitCreateTreeParamsTree[] = [];
+  for (const path in changes) {
+    if (!path) continue;
```
Comment (chingor13) [inline src/github-handler/commit-and-push-handler.ts:28]: When would path be unset?

File: src/github-handler/commit-and-push-handler.ts [hunk 3]
```diff
@@ -0,0 +1,187 @@
...
+/**
+ * Generate and return a GitHub tree object structure
+ * containing the target change data
+ * See https://developer.github.com/v3/git/trees/#tree-object
+ * @param {Changes} changes the set of repository changes
+ * @returns {GitCreateTreeParamsTree[]} The new GitHub changes
+ */
+function genTreeObj(changes: Changes): GitCreateTreeParamsTree[] {
+  const tree: GitCreateTreeParamsTree[] = [];
+  for (const path in changes) {
```
Comment (chingor13) [inline src/github-handler/commit-and-push-handler.ts:27]: If we switch to a built-in `Map` type, this would be something like `for (const path in changes.keys())`

File: src/github-handler/commit-and-push-handler.ts [hunk 4]
```diff
@@ -0,0 +1,187 @@
...
+    if (!path) continue;
+    if (changes[path].content) {
```
Comment (chingor13) [inline src/github-handler/commit-and-push-handler.ts:29]: Is `""` string truthy? If I want an empty string (a.k.a. an empty file), will this instead delete the file?

File: src/github-handler/commit-and-push-handler.ts [hunk 5]
```diff
@@ -0,0 +1,187 @@
...
+ * @param {RepoDomain} origin the the remote repository to push changes to
+ * @param {string} refHEAD the base of the new commit(s)
+ * @param {GitCreateTreeParamsTree[]} tree the set of GitHub changes to upload
+ * @returns {Promise<string>} the GitHub tree SHA
+ */
+async function createTree(
+  logger: Logger,
+  octokit: Octokit,
+  origin: RepoDomain,
+  refHEAD: string,
```
Comment (chingor13) [inline src/github-handler/commit-and-push-handler.ts:64]: nit:
```suggestion
  refHead: string,
```

File: src/github-handler/commit-and-push-handler.ts [hunk 6]
```diff
@@ -0,0 +1,187 @@
...
+import {Octokit} from '@octokit/rest';
+
+/**
+ * Generate and return a GitHub tree object structure
+ * containing the target change data
+ * See https://developer.github.com/v3/git/trees/#tree-object
+ * @param {Changes} changes the set of repository changes
+ * @returns {GitCreateTreeParamsTree[]} The new GitHub changes
+ */
+function genTreeObj(changes: Changes): GitCreateTreeParamsTree[] {
```
Comment (chingor13) [inline src/github-handler/commit-and-push-handler.ts:25]: nit: generally, we don't want to truncate words
```suggestion
function generateTreeObjects(changes: Changes): GitCreateTreeParamsTree[] {
```

File: src/types/index.ts [hunk 7]
```diff
@@ -15,9 +15,27 @@
...
+declare interface GitCreateTreeParamsTree {
+  path?: string;
+  mode?: FileMode;
+  type?: 'blob' | 'tree' | 'commit';
+  sha?: string | null;
+  content?: string;
+}
+
+interface FileData {
+  mode: FileMode;
```
Comment (chingor13) [inline src/types/index.ts:32]: Do we expect all change sources to include the file mode or should it be optional with a default value?

File: src/types/index.ts [hunk 8]
```diff
@@ -15,9 +15,27 @@
...
 
+type FileMode = '100644' | '100755' | '040000' | '160000' | '120000';
+
+/**
+ * GitHub definition of tree
+ */
+declare interface GitCreateTreeParamsTree {
+  path?: string;
+  mode?: FileMode;
+  type?: 'blob' | 'tree' | 'commit';
```
Comment (chingor13) [inline src/types/index.ts:26]: Looks like these are always present?

**Comment by chingor13** [comment]
In general, we still want to camel case abbreviations (e.g. headSha, somethingHead)

File: src/types/index.ts (hunk 7 above)
Comment (TomKristie) [reply src/types/index.ts:32]: I think it's a great idea. The only downside I can see is that we will diverge from the Octokit/GitHub v3 api design where there is no default value. However users don't need to know that we use Octokit - just our interface, so there shouldn't be a problem.

File: src/github-handler/commit-and-push-handler.ts (hunk 5 above)
Comment (TomKristie) [reply src/github-handler/commit-and-push-handler.ts:69]: Just saw your final comment below. Updating

File: test/commit-and-push.ts [hunk 9]
```diff
@@ -0,0 +1,426 @@
...
+        owner: origin.owner,
+        repo: origin.repo,
+        sha: createCommitResponse.data.sha,
+        ref: 'heads/test-branch-name',
+      });
+      // restore
+      stubGetCommit.restore();
+      stubCreateTree.restore();
+      stubCreateCommit.restore();
+      stubUpdateRef.restore();
```
Comment (chingor13) [inline test/commit-and-push.ts:282]: Not entirely sure how sinon stubs work - do stubs persist across requests? If a test fails, will the stub be stuck for the rest of the tests?

Comment (chingor13) [reply test/commit-and-push.ts:302]: A common pattern is to collect a list of stubs to restore and in the teardown block, iterate over all of those registered stubs to clear them out. Alternatively, their might be a sinon method for clearing all stubs?

File: src/github-handler/commit-and-push-handler.ts [hunk 10]
```diff
@@ -0,0 +1,198 @@
...
+ * and resolves with the new tree SHA.
+ * Rejects if GitHub V3 API fails with the GitHub error response
+ * @param {Octokit} octokit The authenticated octokit instance
+ * @param {RepoDomain} origin the the remote repository to push changes to
+ * @param {string} refHead the base of the new commit(s)
+ * @param {GitCreateTreeParamsTree[]} tree the set of GitHub changes to upload
+ * @returns {Promise<string>} the GitHub tree SHA
+ */
+async function createTree(
+  logger: Logger,
```
[the remark this answers is not in the document]
Comment (chingor13) [reply src/github-handler/commit-and-push-handler.ts:66]: This is non-blocking on this PR, but something to consider before 1.0.0 as it probably affects the interface of all the methods we are exporting

File: test/commit-and-push.ts (hunk 9 above)
Comment (TomKristie) [reply test/commit-and-push.ts:302]: According to [best practices](https://semaphoreci.com/community/tutorials/best-practices-for-spies-stubs-and-mocks-in-sinon-js) and some other sites, I have seen wrapping tests with `sinon.test()` and using `sandboxes`, but neither seem to work on my environment so I am using your former suggestion. I've pushed another commit using global stubs and restoring them before each test. It's more manual/explicit since I understand loops are frowned upon in tests.

Comment (TomKristie) [reply test/commit-and-push.ts:302]: i.e. the following sandbox implementation did not work
```
const sandbox = sinon.createSandbox();
describe('test', () => {
  afterEach(() => {
    sandbox.restore();
  });
  it('tests octokit', async() => {
    sandbox.stub(octokit.git, 'createCommit').resolves(createCommitResponse);
  });
}
```
fails if I repeat this test suite.

Comment (TomKristie) [reply test/commit-and-push.ts:302]: I toyed around with it more and got the tests to reset after each test run. The docs are outdated (the method signatures didn't work as in the examples)

File: test/commit-and-push.ts [hunk 11]
```diff
@@ -0,0 +1,392 @@
...
+});
+
+describe('Push', async () => {
+  const changes: Changes = new Map();
+  changes.set('a/foo.txt', new FileData('Foo content'));
+  changes.set('b/bar.txt', new FileData(null));
+  changes.set('baz.exe', new FileData(null, '100755'));
+  changes.set('empty.txt', new FileData(''));
+
+  const tree: TreeObject[] = [
```
Comment (chingor13) [inline test/commit-and-push.ts:33]: Unless these are used across multiple tests within the `describe` block, but it into the test that uses this test data.

File: test/commit-and-push.ts [hunk 12]
```diff
@@ -0,0 +1,392 @@
...
+      mode: '100755',
+      type: 'blob',
+      sha: null,
+    },
+    {
+      path: 'empty.txt',
+      mode: '100644',
+      type: 'blob',
+      content: '',
+    },
```
Comment (chingor13) [inline test/commit-and-push.ts:57]: Since you're exporting `generateTreeObjects`, I might separate each test/edge case for the input file -> `TreeObject`. That way it's more explicit that a `FileData` with null|empty content should create certain `TreeObject`s.

Otherwise, you're only testing then entire `createTree` method which could fail for multiple reasons (and makes it harder to debug). But say that both the `createTree` test and a test for null content fails, then it's easier to deduce that the null content bug is failing both tests.

File: src/github-handler/commit-and-push-handler.ts [hunk 13]
```diff
@@ -0,0 +1,198 @@
...
+  }
+}
+
+export {
+  commitAndPush,
+  createCommit,
+  generateTreeObjects,
+  createTree,
+  updateRef,
+};
```
Comment (chingor13) [inline src/github-handler/commit-and-push-handler.ts:198]: To consider: let's think about what we truly want/need to export (we may need to for testing individual functions)

File: test/commit-and-push.ts [hunk 14]
```diff
@@ -0,0 +1,430 @@
...
+  });
+});
+
+describe('Update branch reference', () => {
+  beforeEach(() => {
+    stubGetCommit.restore();
+    stubCreateTree.restore();
+    stubCreateCommit.restore();
+    stubUpdateRef.restore();
+  });
```
[the remark this answers is not in the document]
Comment (TomKristie) [reply test/commit-and-push.ts]: This is an outdated commit and there's no longer a globally set initial stub. All the stubs are local now. The new behaviour is making sure each test has a clean runtime environment by doing `sinon.restore()`
